src/V3DfgAstToDfg.cpp

```diff
@@ -0,0 +1,239 @@
+    // NODE STATE
+
+    // AstNode::user1p   // DfgVertex for this AstNode
+    const VNUser1InUse m_user1InUse;
+
+    // STATE
+
+    DfgGraph* const m_dfgp;  // The graph being built
+    V3DfgOptimizationContext& m_ctx;  // The optimization context for stats
+    bool m_foundUnhandled = false;  // Found node not implemented as DFG or not implemented 'visit'
+    std::vector<DfgVertex*> m_uncommittedVertices;  // Vertices that we might decide to revert
+
+    // METHODS
+
+    VL_DEBUG_FUNC;
+
+    void markReferenced(AstNode* nodep) {
+        nodep->foreach<AstVarRef>([this](const AstVarRef* refp) {  //
+            getNet(refp->varp())->setHasModRefs();
+        });
+    }
+
+    void commitVertices() { m_uncommittedVertices.clear(); }
+
+    void revertUncommittedVertices() {
+        for (DfgVertex* const vtxp : m_uncommittedVertices) vtxp->unlinkDelete(*m_dfgp);
+        m_uncommittedVertices.clear();
+    }
+
+    DfgVar* getNet(AstVar* varp) {
+        if (!varp->user1p()) {
+            // Note DfgVar vertices are not added to m_uncommittedVertices, because we want to
+            // hold onto them via AstVar::user1p, and the AstVar which might be referenced via
+            // multiple AstVarRef instances, so we will never revert a DfgVar once created. This
+            // means we can end up with DfgVar vertices in the graph which have no connections at
+            // all (which is fine for later processing).
+            varp->user1p(new DfgVar{*m_dfgp, varp});
+        }
+        return varp->user1u().to<DfgVar*>();
+    }
+
+    DfgVertex* getVertex(AstNode* nodep) {
+        DfgVertex* vtxp = nodep->user1u().to<DfgVertex*>();
+        UASSERT_OBJ(vtxp, nodep, "Missing Dfg vertex");
+        return vtxp;
+    }
+
+    // VISITORS
+    void visit(AstNode* nodep) override {
+        // Conservatively treat this node as unhandled
+        m_foundUnhandled = true;
+        markReferenced(nodep);
+    }
+    void visit(AstCell* nodep) override { markReferenced(nodep); }
+    void visit(AstNodeProcedure* nodep) override { markReferenced(nodep); }
+    void visit(AstVar* nodep) override {
+        // Mark ports as having external references
+        if (nodep->isIO()) getNet(nodep)->setHasExtRefs();
+        // Mark variables that are the target of a hierarchical reference
+        // (these flags were set up in DataflowPrepVisitor)
+        if (nodep->user2()) getNet(nodep)->setHasExtRefs();
+    }
+
+    void visit(AstAssignW* nodep) override {
+        // Explicitly turned off optimization
+        if (!nodep->fileline()->optDfgOn()) {
+            markReferenced(nodep);
+            return;
+        }
+
+        // Cannot handle assignments between different types at the moment
+        if (!nodep->lhsp()->dtypep()->skipRefp()->similarDType(
+                nodep->rhsp()->dtypep()->skipRefp())) {
+            markReferenced(nodep);
+            ++m_ctx.m_nonRepresentableDType;
+            return;
+        }
```

Comment:
Is there a more robust way of detecting if the two sides are simple packed types (all I really care about is that they are both packed and have the same width, possibly signedness)? This for example fails if one side is `logic` but the other is `logic [0:0]` (which you get from parametric source), even though those two are fine here.

Reply:
Not currently, the type handling is messy.  Note "logic" and "logic [0]" while in your code and much of the code can be handled similarly they are not the same language wise, or if hits DPI or tracing.